### PR TITLE
Disallow untyped calls

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

**Comment by danieleades** [comment]
```
/opt/hostedtoolcache/Python/3.12.4/x64/lib/python3.12/site-packages/sphinx/locale/__init__.py:docstring of sphinx.locale.init:1: WARNING: py:class reference target not found: Path [ref.class]
```

i seem to run into these same bugs everywhere i go!

File: pyproject.toml [hunk 1]
```diff
@@ -279,6 +280,7 @@ module = [
 ]
 ignore_missing_imports = true
 
+
```
Comment (picnixz) [inline pyproject.toml]: unrelated change?

```suggestion
```

File: tests/test_builders/test_build_html_5_output.py [hunk 2]
```diff
@@ -1,17 +1,20 @@
 """Test the HTML builder and check output against XPath."""
 
 import re
+from collections.abc import Callable, Iterable
+from typing import Literal
+from xml.etree.ElementTree import Element
```
Comment (picnixz) [inline tests/test_builders/test_build_html_5_output.py]: Put them in a TYPE_CHECKING block instead.

File: tests/test_builders/test_build_linkcheck.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 from base64 import b64encode
 from http.server import BaseHTTPRequestHandler
 from queue import Queue
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any
```
Comment (picnixz) [inline tests/test_builders/test_build_linkcheck.py]: Ditto.

File: tests/test_builders/test_build_linkcheck.py [hunk 4]
```diff
@@ -436,7 +438,7 @@ def do_GET(self):
     assert row['status'] == 'ignored'
 
 
-def custom_handler(valid_credentials=(), success_criteria=lambda _: True):
+def custom_handler(valid_credentials: tuple[str, str] | None = None, success_criteria: Callable[[Any], bool] = lambda _: True) -> type[BaseHTTPRequestHandler]:
```
Comment (picnixz) [inline tests/test_builders/test_build_linkcheck.py]: Split the signature.

File: tests/test_builders/test_build_text.py [hunk 5]
```diff
@@ -1,12 +1,14 @@
 """Test the build process with Text builder with the test root."""
 
+from typing import Any
```
Comment (picnixz) [inline tests/test_builders/test_build_text.py]: In a TYPE_CHECKING block (I won't mark the others, but remember to check that).

File: tests/test_util/test_util_inventory.py [hunk 6]
```diff
@@ -1,7 +1,10 @@
 """Test inventory util functions."""
+from __future__ import annotations
```
Comment (picnixz) [inline tests/test_util/test_util_inventory.py:3]: And don't forget this one in other test files.

**Review by picnixz — COMMENTED** [review]
I assume you want to fix the missing annotations upstream as well?

**Comment by picnixz** [comment]
> i seem to run into these same bugs everywhere i go!

This is definitely because of the `__future__` annotations I think. But ... it might also be because of the fact that `Path` is not documented at all actually... (or that it's real module is not `pathlib`). I haven't touched the pathlib part yet in CPython but maybe there was some changes out there.

File: tests/test_builders/test_build_linkcheck.py (hunk 4 above)
Comment (danieleades) [reply tests/test_builders/test_build_linkcheck.py]: what do you mean?

Comment (picnixz) [reply tests/test_builders/test_build_linkcheck.py]: I meant to be on two lines (but maybe it's not that long, it might be the GitHub UI that made me felt that way).

**Comment by danieleades** [comment]
> I assume you want to fix the missing annotations upstream as well?

at some point, if i find the time

File: tests/test_builders/test_build_html_5_output.py (hunk 2 above)
Comment (danieleades) [reply tests/test_builders/test_build_html_5_output.py]: for reasons that are not obvious to me, ruff complains if these are moved into a type checking block

> Move import {IMPORT} out of type-checking block. Import is used for more than type hinting. [Ruff [TCH004](https://docs.astral.sh/ruff/rules/runtime-import-in-type-checking-block)]

File: tests/test_util/test_util_inventory.py (hunk 6 above)
Comment (danieleades) [reply tests/test_util/test_util_inventory.py:3]: ruff should lint any files that use modern type syntax and forget to add `from __future__ import annotations`

File: tests/test_builders/test_build_html_5_output.py (hunk 2 above)
Comment (picnixz) [reply tests/test_builders/test_build_html_5_output.py]: It's because `from future import __annotations__` is missing.

File: tests/test_builders/test_build_linkcheck.py [hunk 7]
```diff
@@ -843,7 +849,7 @@ def test_TooManyRedirects_on_HEAD(app, monkeypatch):
     }
 
 
-def make_retry_after_handler(responses):
+def make_retry_after_handler(responses: list[tuple[int, str | None]]) -> type[BaseHTTPRequestHandler]:
```
Comment (jayaddison) [inline tests/test_builders/test_build_linkcheck.py:852]: Probably technically an `Iterable[tuple[int, str | None]]` instead of a list here (a series of responses that can be emitted in sequence, although not all of them may be consumed).

File: sphinx/builders/_epub_base.py [hunk 8]
```diff
@@ -108,8 +108,8 @@ class NavPoint(NamedTuple):
 
 def sphinx_smarty_pants(t: str, language: str = 'en') -> str:
     t = t.replace('&quot;', '"')
-    t = smartquotes.educateDashesOldSchool(t)
-    t = smartquotes.educateQuotes(t, language)
+    t = smartquotes.educateDashesOldSchool(t)  # type: ignore[no-untyped-call]
```
Comment (jayaddison) [inline sphinx/builders/_epub_base.py:111]: Could we try ignoring `untyped-call` on a [per-module](https://mypy.readthedocs.io/en/stable/config_file.html#per-module-and-global-options) basis?  (it seems like we're not quite ready for that check on these calls)

**Review by AA-Turner — APPROVED** [review]
